_posts/2020-12-14-how-to-lock-down-csp-using-swachbuckle.md

```diff
@@ -0,0 +1,224 @@
+
+```csharp
+// Startup.cs
+public void ConfigureServices()
+{
+    services
+        .AddOptions<SwaggerUIOptions>()
+        .Configure<IHttpContextAccessor>((swaggerUiOptions, httpContextAccessor) =>
+        {
+            // 1. Take a reference of the original Stream factory which reads from Swashbuckle's embedded resources
+            var originalIndexStreamFactory = options.IndexStream;
+
+            // 2. Override the Stream factory
+            options.IndexStream = () =>
+            {
+                // 3. Read the original index.html file
+                using var originalStream = originalIndexStreamFactory();
+                using var originalStreamReader = new StreamReader(originalStream);
+                var originalIndexHtmlContents = originalStreamReader.ReadToEnd();
+
+                // 4. Get the request-specific nonce generated by NetEscapades.AspNetCore.SecurityHeaders
+                var requestSpecificNonce = httpContextAccessor.HttpContext.GetNonce();
+
+                // 5. Replace inline `<script>` and `<style>` tags by adding a `nonce` attribute to them
+                var nonceEnabledIndexHtmlContents = originalIndexHtmlContents
+                    .Replace("<script>", $"<script nonce=\"{requestSpecificNonce}\">", StringComparison.OrdinalIgnoreCase)
+                    .Replace("<style>", $"<style nonce=\"{requestSpecificNonce}\">", StringComparison.OrdinalIgnoreCase);
+
+                // 6. Return a new Stream that contains our modified contents
+                return new MemoryStream(Encoding.UTF8.GetBytes(nonceEnabledIndexHtmlContents));
+            };
+        });
+}
+```
+
+The last bit of configuration is to allow nonces for styles and scripts in our CSP:
+
+```csharp
+public void Configure()
+{
+    // Note this sample only focuses on CSP, but you might want to set other security-related headers if you use this library
+    app.UseSecurityHeaders(policyCollection =>
+    {
+        policyCollection.AddContentSecurityPolicy(csp =>
+        {
+            // Only allow loading resources from this app by default
+            csp.AddDefaultSrc().Self();
+
+            csp.AddStyleSrc()
+               .Self()
+               // Allow nonce-enabled <style> tags
+               .WithNonce();
+
+            csp.AddScriptSrc()
+               .Self()
+               // Allow nonce-enabled <script> tags
+               .WithNonce();
+
+            [...]
+        });
+    });
+}
+```
+
+If we access our swagger-ui page again, everything seems to be working fine! 🎉
+Refreshing the page while taking a look at the HTML code, we can see that the `nonce` attribute gets generated on every request.
+
+If we look at the DevTools console again, we can see two errors remain related to loading images through the `data` protocol.
+From what I can see, they're related to the Swagger logo in the top left corner and the down-facing arrow in the definition dropdown in the top-right corner.
```

Comment:
```suggestion
If we open our browser developer tools, the console should show two errors related to loading images through the `data` protocol.
Because this protocol is not allowed in our content security policy, the browser refuses to load them.
From what I can see, they're related to the Swagger logo in the top-left corner and the down-facing arrow in the definition dropdown in the top-right corner.
```